### PR TITLE
ObjectId and Decimal128

### PR DESCRIPTION
**Pull request by ironage**
Initial support. Requires most recent core/sync on core-6.1 branches.

### CODE REVIEW

File: src/object_accessor.hpp [hunk 1]
```diff
@@ -158,6 +158,8 @@ ValueType Object::get_property_value_impl(ContextType& ctx, const Property &prop
         case PropertyType::String: return ctx.box(m_obj.get<StringData>(column));
         case PropertyType::Data:   return ctx.box(m_obj.get<BinaryData>(column));
         case PropertyType::Date:   return ctx.box(m_obj.get<Timestamp>(column));
+        case PropertyType::ObjectId: return ctx.box(m_obj.get<ObjectId>(column));
+        case PropertyType::Decimal:  return ctx.box(m_obj.get<Decimal>(column));
```
Comment (kneth) [inline src/object_accessor.hpp]: Why not `Decimal128`?

File: src/property.hpp [hunk 2]
```diff
@@ -51,6 +53,9 @@ enum class PropertyType : unsigned char {
     // deprecated and remains only for reading old files
     Any    = 9,
 
+    ObjectId = 10,
+    Decimal = 11,
```
Comment (kneth) [inline src/property.hpp]: Why not `Decimal128`?

Comment (jedelbo) [reply src/property.hpp]: We don't have Int64 and Int32.

Comment (kneth) [reply src/property.hpp]: So why does the PR's title contain `Decimal128`?

Comment (ironage) [reply src/property.hpp]: As far as I can tell, according to mongo's type [documentation](https://docs.mongodb.com/manual/reference/bson-types/#document-bson-type-date) the standard is to use `Decimal128` for enum type names, but use "decimal" for the string type description name. I can change the property type enum according to that.

**Comment by jedelbo** [comment]
This PR should be rebased on core-6.1 branch

**Comment by ironage** [comment]
Rebased. This will require a version of core with https://github.com/realm/realm-core/pull/3601 to link properly.

**Comment by cmelchior** [comment]
I guess this branch should be redirected against the v10 branch now?

**Comment by ironage** [comment]
Redirected to v10.
The core version that this requires (containing realm/realm-core#3601) is now available - but it fails to build with sync due to the c++17 changes that come along with it. We are now blocked on getting a c++17 version of sync compatible with core v10.0.0-alpha.1

**Comment by ironage** [comment]
Since a large portion of the original part of this PR was already merged, the remaining work is about adding proper support for Optional<ObjectId> and improving the tests around that.

This now compiles and links correctly with the new sync version provided by the underlying v10 branch. There are however, unrelated segfaults and hangs in the sync tests on v10 so CI won't be clean.

File: src/results.cpp [hunk 3]
```diff
@@ -1037,6 +1037,7 @@ REALM_RESULTS_TYPE(util::Optional<bool>)
 REALM_RESULTS_TYPE(util::Optional<int64_t>)
 REALM_RESULTS_TYPE(util::Optional<float>)
 REALM_RESULTS_TYPE(util::Optional<double>)
+REALM_RESULTS_TYPE(util::Optional<ObjectId>)
```
Comment (kneth) [inline src/results.cpp]: I don't see `REALM_RESULTS_TYPE(util::Optional<Decimal128>)`.

File: src/impl/object_accessor_impl.hpp [hunk 4]
```diff
@@ -112,6 +112,7 @@ class CppContext {
     util::Any box(util::Optional<double> v) const { return v; }
     util::Any box(util::Optional<float> v) const { return v; }
     util::Any box(util::Optional<int64_t> v) const { return v; }
+    util::Any box(util::Optional<ObjectId> v) const { return v; }
```
Comment (kneth) [inline src/impl/object_accessor_impl.hpp]: I don't see `util::Any box(util::Optional<Decimal128> v) const { return v; }`

File: src/list.cpp [hunk 5]
```diff
@@ -482,6 +482,7 @@ REALM_PRIMITIVE_LIST_TYPE(util::Optional<bool>)
 REALM_PRIMITIVE_LIST_TYPE(util::Optional<int64_t>)
 REALM_PRIMITIVE_LIST_TYPE(util::Optional<float>)
 REALM_PRIMITIVE_LIST_TYPE(util::Optional<double>)
+REALM_PRIMITIVE_LIST_TYPE(util::Optional<ObjectId>)
```
Comment (kneth) [inline src/list.cpp]: I don't see `REALM_PRIMITIVE_LIST_TYPE(util::Optional<Decimal128>)`

File: src/results.cpp (hunk 3 above)
Comment (tgoyne) [reply src/results.cpp]: `Decimal128` can store `null` internally like StringData etc. while `ObjectId` can't.

**Pull request merged**